server/game/gameSystems/MoveCardSystem.ts

```diff
@@ -81,19 +108,35 @@ export class MoveCardSystem<TContext extends AbilityContext = AbilityContext> ex
         event.switchTarget = properties.switchTarget;
         event.changePlayer = properties.changePlayer;
         event.destination = properties.destination;
-        event.bottom = properties.bottom;
         event.shuffle = properties.shuffle;
+        event.locationFilter = properties.locationFilter;
+        event.options = { bottom: !!properties.bottom };
     }
 
-    public override canAffect(card: Card, context: TContext, additionalProperties = {}): boolean {
-        const { changePlayer, destination } = this.generatePropertiesFromContext(context, additionalProperties) as IMoveCardProperties;
-        return (
-            (!changePlayer ||
-              (!card.hasRestriction(EffectName.TakeControl, context) &&
-                !card.anotherUniqueInPlay(context.player))) &&
-                (!destination || context.player.isLegalLocationForCardType(card.type, destination)) &&
-                !EnumHelpers.isArena(card.location) &&
-                super.canAffect(card, context)
-        );
+    public override canAffect(card: Card, context: TContext, additionalProperties = {}, mustChangeGameState = GameStateChangeRequired.None): boolean {
+        const { changePlayer, destination, locationFilter } = this.generatePropertiesFromContext(context, additionalProperties) as IMoveCardProperties;
+
+        // Ensure that we have a valid destination and that the card can be moved there
+        if (!destination || !context.player.isLegalLocationForCardType(card.type, destination)) {
+            return false;
+        }
+
+        // Ensure that the card has no restrictions that would prevent it from being moved to a different player
+        if (changePlayer && (card.hasRestriction(EffectName.TakeControl, context) || card.anotherUniqueInPlay(context.player))) {
+            return false;
+        }
+
+        // Ensure that if the card is in play, it can only be moved to the hand or deck
+        if (EnumHelpers.isArena(card.location) && destination !== Location.Hand && destination !== Location.Deck) {
```

Comment:
Looking at these again, I'm kind of conflicted as to whether these should silently fail or actually throw a contract fail error. I generally lean towards the latter so that we can catch static configuration glitches early, unless there's a strong case to be made for why it should be allowable to attempt to perform an action that fails (which we do sometimes). Just can't think of a strong reason to here.